src/query-builder/QueryBuilder.ts

```diff
@@ -993,51 +993,68 @@ export abstract class QueryBuilder<Entity> {
 
             if (metadata.hasRelationWithPropertyPath(path)) {
                 const relation = metadata.findRelationWithPropertyPath(path)!;
+                const isEntityTarget = relation.inverseEntityMetadata.target === entity[key].constructor;
 
-                // There's also cases where we don't want to return back all of the properties.
-                // These handles the situation where someone passes the model & we don't need to make
-                // a HUGE `where` to uniquely look up the entity.
+                const primaryColumns = relation.inverseEntityMetadata.primaryColumns;
+                const seenPrimaryColumns = new Set<string>();
+                for (const primaryColumn of primaryColumns) {
+                    const entityValueMap = primaryColumn.getEntityValueMap(entity[key]);
+                    for (const [columnName, value] of Object.entries(entityValueMap || {})) {
+                        if (value) {
+                            seenPrimaryColumns.add(columnName);
+                        } else {
+                            break;
+                        }
+                    }
+                }
+                const hasOnlyPrimaryKeys = primaryColumns.length > 0 && seenPrimaryColumns.size === Object.keys(entity[key]).length;
```

Comment:
If it's only primary columns do we actually care if it hits the optimizations?

In theory, it would be the same either way, wouldn't it?

Reply:
The update flow uses this method and fails if you pass `contact.id` vs `contactId`. 
ex. https://github.com/typeorm/typeorm/pull/8005#issuecomment-891214828
That's the reason I added the hasOnlyPrimaryKeys check back.